singularity-opac/photons/mean_opacity_photons.hpp

```diff
@@ -0,0 +1,209 @@
+// Department of Energy/National Nuclear Security Administration. All
+// rights in the program are reserved by Triad National Security, LLC,
+// and the U.S. Department of Energy/National Nuclear Security
+// Administration. The Government is granted for itself and others
+// acting on its behalf a nonexclusive, paid-up, irrevocable worldwide
+// license in this material to reproduce, prepare derivative works,
+// distribute copies to the public, perform publicly and display
+// publicly, and to permit others to do so.
+// ======================================================================
+
+#ifndef SINGULARITY_OPAC_PHOTONS_MEAN_OPACITY_PHOTONS_
+#define SINGULARITY_OPAC_PHOTONS_MEAN_OPACITY_PHOTONS_
+
+#include <cmath>
+
+#include <ports-of-call/portability.hpp>
+#include <singularity-opac/base/radiation_types.hpp>
+#include <singularity-opac/base/robust_utils.hpp>
+#include <singularity-opac/base/sp5.hpp>
+#include <singularity-opac/constants/constants.hpp>
+#include <spiner/databox.hpp>
+
+#include <singularity-opac/photons/mean_photon_variant.hpp>
+
+namespace singularity {
+namespace photons {
+namespace impl {
+
+#define EPS (10.0 * std::numeric_limits<Real>::min())
+
+// TODO(BRR) Note: It is assumed that lambda is constant for all densities,
+// temperatures, and Ye
+
+template <typename pc = PhysicalConstantsCGS>
+class MeanOpacity {
+
+ public:
+  MeanOpacity() = default;
+  template <typename Opacity>
+  MeanOpacity(const Opacity &opac, const Real lRhoMin, const Real lRhoMax,
+              const int NRho, const Real lTMin, const Real lTMax, const int NT,
+              Real *lambda = nullptr) {
+    lkappaPlanck_.resize(NRho, NT);
+    lkappaPlanck_.setRange(0, lTMin, lTMax, NT);
+    lkappaPlanck_.setRange(1, lRhoMin, lRhoMax, NRho);
+    lkappaRosseland_.copyMetadata(lkappaPlanck_);
+
+    // Fill tables
+    for (int iRho = 0; iRho < NRho; ++iRho) {
+      Real lRho = lkappaPlanck_.range(1).x(iRho);
+      Real rho = fromLog_(lRho);
+      for (int iT = 0; iT < NT; ++iT) {
+        Real lT = lkappaPlanck_.range(0).x(iT);
+        Real T = fromLog_(lT);
+        Real kappaPlanckNum = 0.;
+        Real kappaPlanckDenom = 0.;
+        Real kappaRosselandNum = 0.;
+        Real kappaRosselandDenom = 0.;
+        // Integrate over frequency
+        const int nnu = 100;
+        const Real lnuMin = toLog_(1.e-3 * pc::kb * fromLog_(lTMin) / pc::h);
+        const Real lnuMax = toLog_(1.e3 * pc::kb * fromLog_(lTMax) / pc::h);
+        const Real dlnu = (lnuMax - lnuMin) / (nnu - 1);
+        for (int inu = 0; inu < nnu; ++inu) {
+          const Real lnu = lnuMin + inu * dlnu;
+          const Real nu = fromLog_(lnu);
+          kappaPlanckNum += opac.AbsorptionCoefficient(rho, T, nu, lambda) /
+                            rho * opac.ThermalDistributionOfTNu(T, nu) * nu *
+                            dlnu;
+          kappaPlanckDenom += opac.ThermalDistributionOfTNu(T, nu) * nu * dlnu;
```

Comment:
If we moved to a fast log would that cause any problems here? I think it might mean decoupling the logs used in the integration from the logs used in the gridding.

That's a problem for future us, but I'm curious on your thoughts.

Reply:
integrating over `d log nu` here requires `d log nu = d nu / nu` which I assume isn't satisfied by fast log. This is also only done during initialization so accelerating this may not be necessary.

Reply:
yeah just thinking that fast log gridding would mean two separate log methods, one for log, one for fast log